### PR TITLE
Preparatory work for fractures + wells

### PR DESCRIPTION
**Pull request by paveltomin**
For now, assuming wells are not directly connected to the fractures.

In this version, only `SinglePhasePoromechanicsConformingFracturesReservoir` solver can run, although I am having hard time to build a well-behaving example. Other versions of solvers with wells are drafted but commented out for now.

Added poromechanics initialization for contact solver - technically runs but needs refinements.

Side thing: add default template types to avoid too long nested definitions

### CODE REVIEW

File: src/coreComponents/physicsSolvers/multiphysics/PoromechanicsSolver.hpp [hunk 1]
```diff
@@ -212,6 +212,18 @@ class PoromechanicsSolver : public CoupledSolver< FLOW_SOLVER, MECHANICS_SOLVER
     m_performStressInitialization = performStressInitialization;
   }
 
+  virtual bool updateConfiguration( DomainPartition & domain ) override
+  {
+    if( m_performStressInitialization ) // assume no configuration change during initialization step
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/multiphysics/PoromechanicsSolver.hpp]: @CusiniM @castelletto1 @Guotong-Ren @jhuang2601 please let me know if this makes sense...

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/multiphysics/PoromechanicsSolver.hpp]: i decided to remove that change

**Comment by paveltomin** [comment]
@CusiniM are you ok to move this forward?

**Comment by paveltomin** [comment]
this PR is relatively limited now, please review

**Comment by paveltomin** [comment]
review?

**Review by jhuang2601 — APPROVED** [review]
Thanks for enabling fault initialization

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp [hunk 2]
```diff
@@ -305,12 +305,12 @@ void SinglePhaseFVM< BASE >::applySystemSolution( DofManager const & dofManager,
   } );
 }
 
-template< >
-void SinglePhaseFVM< SinglePhaseBase >::assembleFluxTerms( real64 const dt,
```
Comment (Guotong-Ren) [inline src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp:309]: I am trying to understand the logic here. what is proposed is always using BASE besides singlePhaseProppantBase specialization? And I guess this implementation is assumed not only being restricted to SinglePhaseBase.

**Comment by Guotong-Ren** [comment]
this question is not quite relevant to this PR. I am curious how much work do you estimate to prototype when well is connected to fractures? need adding extra blocks in the Jacobin matrix?

**Comment by paveltomin** [comment]
> this question is not quite relevant to this PR. I am curious how much work do you estimate to prototype when well is connected to fractures? need adding extra blocks in the Jacobin matrix?

I would redirect this to @CusiniM for better estimate.
The code pieces should be there, but I am not exactly sure how to connect them.

**Comment by paveltomin** [comment]
> > this question is not quite relevant to this PR. I am curious how much work do you estimate to prototype when well is connected to fractures? need adding extra blocks in the Jacobin matrix?
> 
> I would redirect this to @CusiniM for better estimate. The code pieces should be there, but I am not exactly sure how to connect them.

I mean, I am ready to work on that, just need some guidance where to start.

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp (hunk 2 above)
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp:309]: If I remember correctly, the change should be just related to the default template argument (`SinglePhaseBase`), no changes in other logic. `SinglePhaseFVM< SinglePhaseProppantBase >::assembleFluxTerms` is still down below.

**Comment by CusiniM** [comment]
> > > this question is not quite relevant to this PR. I am curious how much work do you estimate to prototype when well is connected to fractures? need adding extra blocks in the Jacobin matrix?
> > 
> > 
> > I would redirect this to @CusiniM for better estimate. The code pieces should be there, but I am not exactly sure how to connect them.
> 
> I mean, I am ready to work on that, just need some guidance where to start.

I don't think it's a lot of work. However, it would be nice to first refactor the way wells are handled by introducing a `WellStencil` (or something similar)  and then use a dedicated stencil for those connections. Once the connections are well defined adding the terms to the jacobian is trivial IMO.

File: src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanics.cpp [hunk 3]
```diff
@@ -322,16 +322,18 @@ void SinglePhasePoromechanics< FLOW_SOLVER, MECHANICS_SOLVER >::updateBulkDensit
                                                subRegion );
 }
 
-template class SinglePhasePoromechanics< SinglePhaseBase >;
+template class SinglePhasePoromechanics<>;
 template class SinglePhasePoromechanics< SinglePhaseBase, SolidMechanicsLagrangeContact >;
 template class SinglePhasePoromechanics< SinglePhaseBase, SolidMechanicsEmbeddedFractures >;
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanics.cpp:327]: is this necessary? Isn't it already being instantiated in `SinglePhasePoromechanicsConformingFractures.cpp` ?

Comment (CusiniM) [reply src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanics.cpp:327]: it also seems you are not registering it here.

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanics.cpp:327]: I remember that somehow it was important but I can double-check
Would it miss versions of SinglePhasePoromechanics functions when only present in SinglePhasePoromechanicsConformingFractures?

Comment (CusiniM) [reply src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanics.cpp:327]: ah yes, sounds possible

**Pull request merged**